### PR TITLE
Fractional seek fix

### PR DESCRIPTION
**Pull request by daschuer**
Fix a call of the fractional seek function with a sample value. 
Rename some functions and parameters to avoid this error in future.

### CODE REVIEW

File: src/engine/enginebuffer.cpp [hunk 1]
```diff
@@ -348,7 +348,7 @@ EngineBuffer::~EngineBuffer() {
 }
 
 double EngineBuffer::fractionalPlayposFromAbsolute(double absolutePlaypos) {
-    double fFractionalPlaypos = 0.0;
+    double fFractionalPlaypos;
```
Comment (ywwg) [inline src/engine/enginebuffer.cpp]: instead, restore default value and remove else condition

File: src/engine/enginebuffer.cpp [hunk 2]
```diff
@@ -595,29 +595,31 @@ void EngineBuffer::slotPassthroughChanged(double enabled) {
 }
 
 // WARNING: This method runs in both the GUI thread and the Engine Thread
-void EngineBuffer::slotControlSeek(double change) {
-    doSeek(change, SEEK_STANDARD);
+void EngineBuffer::slotControlSeek(double fractionapPos) {
```
Comment (ywwg) [inline src/engine/enginebuffer.cpp]: typo.  Please make sure check your code for this kind of basic spelling error!

File: src/engine/enginebuffer.cpp [hunk 3]
```diff
@@ -595,29 +595,31 @@ void EngineBuffer::slotPassthroughChanged(double enabled) {
...
+    doSeekFractional(fractionapPos, SEEK_STANDARD);
 }
 
 // WARNING: This method runs from SyncWorker and Engine Worker
-void EngineBuffer::slotControlSeekAbs(double abs) {
-    doSeek(abs / m_file_length_old, SEEK_STANDARD);
+void EngineBuffer::slotControlSeekAbs(double playPosition) {
+    doSeek(playPosition, SEEK_STANDARD);
```
Comment (ywwg) [inline src/engine/enginebuffer.cpp]: this changes doSeek from seeking to a proportional position to an absolute position.  Please do not change the API because this may break something elsewhere.

**Comment by daschuer** [comment]
Notes addresses.
Now we have doSeekFractional and doSeekPlayPos. Just doSeek was misleading because the engine usually uses playPos elsewhere with seek functions without a suffix.

File: src/engine/enginebuffer.cpp [hunk 4]
```diff
@@ -595,29 +593,31 @@ void EngineBuffer::slotPassthroughChanged(double enabled) {
...
+void EngineBuffer::doSeekFractional(double fractionalPos, enum SeekRequest seekType) {
     // Prevent NaN's from sneaking into the engine.
-    if (isnan(change)) {
+    if (isnan(fractionalPos)) {
         return;
     }
-
     // Find new playpos, restrict to valid ranges.
-    double new_playpos = round(change * m_file_length_old);
+    double new_playpos = round(fractionalPos * m_file_length_old);
```
Comment (ywwg) [inline src/engine/enginebuffer.cpp]: make the adjustment to even-ness here -- we shouldn't make a call with a non-even playposition

File: src/engine/enginebuffer.cpp [hunk 5]
```diff
@@ -1006,8 +1006,8 @@ void EngineBuffer::process(CSAMPLE* pOutput, const int iBufferSize) {
         if ((m_playButton->get() || (m_fwdButton->get() || m_backButton->get()))
                 && end_of_track) {
             if (repeat_enabled) {
-                double seekPosition = at_start ? m_file_length_old : 0;
-                doSeek(seekPosition, SEEK_STANDARD);
+                double fractionapPos = at_start ? 1.0 : 0;
+                doSeekFractional(fractionapPos, SEEK_STANDARD);
```
Comment (ywwg) [inline src/engine/enginebuffer.cpp]: still typo!

File: src/engine/enginebuffer.cpp (hunk 4 above)
Comment (daschuer) [inline src/engine/enginebuffer.cpp]: Forcing to even is already done at to places later in the code.

In a separate PR we should replace the double playpos by integer frames to avoid all these checks. 
In places where sample fractions make scene we should switch to double frames. This way we can express the position of a beat and we can deal with with scaling. 
But currently we cannot do this, since a position of 3 will be ironed to an even sample value. Once we have a position on stereo frames it will work.

**Comment by daschuer** [comment]
Notes addressed!

File: src/engine/enginebuffer.cpp (hunk 4 above)
Comment (ywwg) [inline src/engine/enginebuffer.cpp]: > Forcing to even is already done at to places later in the code.

the multiplication as written can easily result in odd values.

Comment (daschuer) [inline src/engine/enginebuffer.cpp]: yes sure and it is solved here: https://github.com/daschuer/mixxx/blob/fractionalseekfix/src/engine/enginebuffer.cpp#L627

**Comment by daschuer** [comment]
Conflicts, Ready for merge?

File: src/engine/enginebuffer.cpp (hunk 4 above)
Comment (ywwg) [inline src/engine/enginebuffer.cpp]: yes but that type of fix should really be a DEBUG_ASSERT_AND_HANDLE -- It's better to fix even-ness problems where we know they can appear, like here.  Otherwise we are guaranteeing that this function is generating invalid output.  Put the fix here please.

**Comment by daschuer** [comment]
Ok, check is in place and some improvements in some other places.

File: src/engine/enginebuffer.cpp [hunk 6]
```diff
@@ -929,18 +930,13 @@ void EngineBuffer::process(CSAMPLE* pOutput, const int iBufferSize) {
         if (!bCurBufferPaused) {
             // The fileposition should be: (why is this thing a double anyway!?
             // Integer valued.
-            double filepos_play_rounded = round(m_filepos_play);
-            if (filepos_play_rounded != m_filepos_play) {
-                qWarning() << __FILE__ << __LINE__ << "ERROR: filepos_play is not round:" << m_filepos_play;
+            double playFrame = m_filepos_play / kSamplesPerFrame;
+            double filepos_play_rounded = round(playFrame) * kSamplesPerFrame;
```
Comment (ywwg) [inline src/engine/enginebuffer.cpp]: why divide and multiply when you can just do the same mod and add we've been using?  Also this idiom is repeated a few times and probably exists elsewhere in the code.  Factor it out into an inlined function in sample_util called roundToFrame(position);

Comment (daschuer) [inline src/engine/enginebuffer.cpp]: It is a preview of a solution when we move to a playposition in frames. Currently kSamplesPerFrame is const 2. that may change as well. 

It also fixes a minor rounding error. Example a track has 6 Samples and you wish to seek to 58 %
The track was samples at 0% 33% and 66%. 100% is the end position.     
The seek should be done to the 3. frame position since 66% is closest to 58 %.
The desired sample index is 4.

The old code: 
6 * 0,58 = 3,48 
round -> 3
force even = 2

The new code 
6 / 2 * 0,58 = 1,74 
round -> 2
2 * 2 = 4

Comment (daschuer) [inline src/engine/enginebuffer.cpp]: > Factor it out into an inlined function in sample_util called roundToFrame(position);

Yes, a good idea. I just want to keep this PR focused. Can we do it in a separate PR? 
How about changing the api parts to frames to avoid the issue entirely?

Comment (ywwg) [inline src/engine/enginebuffer.cpp]: well this is mostly a cleanup PR and factoring out a small two-liner seems fine and reduces the chance of typos.  No need to change API

Comment (daschuer) [inline src/engine/enginebuffer.cpp]: Does it mean this is merge-able now?

**Comment by ywwg** [comment]
lgtm

**Comment by daschuer** [comment]
Thank you for review!
